Constructions/eps_additon.ml

```diff
@@ -0,0 +1,2136 @@
+    ASSUME `proper_nat_construct a1`;
+    ASSUME `proper_nat_construct b1`] THEN 
+  DISCH_TAC THEN 
+  ASM_REWRITE_TAC[] THEN 
+  IS_EXPR_TYPE_TAC THEN 
+  REPEAT DISCH_TAC THEN 
+  ASM_REWRITE_TAC[] THEN
+  CONSTRUCTION_TO_QUOTE_TAC THEN
+  LAW_OF_DISQUO_TAC THEN 
+  MP_TAC(SPEC `b1:epsilon` (ASSUME 
+    `!y. proper_nat_construct a1
+    ==> proper_nat_construct y
+    ==> (eval (add_ebin a1 y) to (nat)) =
+    add_unary (eval (a1) to (nat)) (eval (y) to (nat))`)) THEN 
+  REWRITE_TAC[
+    y_quoapp_case1_sub;
+    ASSUME `proper_nat_construct a1`;
+    ASSUME `proper_nat_construct b1`] THEN 
+  DISCH_TAC THEN 
+  ASM_REWRITE_TAC[add_odd_even]
+  ;MP_TAC(APP_DISQUO 
+    `QuoConst "thenOne" (TyBiCons "fun" (TyBase "nat") (TyBase "nat"))` 
+    `a1:epsilon`) THEN 
+  MP_TAC(APP_DISQUO 
+    `QuoConst "thenOne" (TyBiCons "fun" (TyBase "nat") (TyBase "nat"))` 
+    `b1:epsilon`) THEN 
+  MP_TAC(APP_DISQUO 
+    `QuoConst "thenZero" (TyBiCons "fun" (TyBase "nat") (TyBase "nat"))` 
+    `(add_ebin (QuoConst "One" (TyBase "nat")) (add_ebin a1 b1)):epsilon`) THEN 
+  PROPER_TYPE_TAC(`a1:epsilon`) THEN 
+  PROPER_TYPE_TAC(`b1:epsilon`) THEN 
+  MP_TAC(SPECL [`a1:epsilon`;`b1:epsilon`] thm1) THEN 
+  REWRITE_TAC[
+    ASSUME `proper_nat_construct a1`;
+    ASSUME `proper_nat_construct b1`] THEN 
+  DISCH_TAC THEN 
+  ASM_REWRITE_TAC[] THEN
+  MP_TAC(SPECL [`a1:epsilon`;`b1:epsilon`] lemma6) THEN  
+  REWRITE_TAC[
+    ASSUME `proper_nat_construct a1`;
+    ASSUME `proper_nat_construct b1`] THEN 
+  DISCH_TAC THEN 
+  MP_TAC(SPECL [
+    `(QuoConst "One" (TyBase "nat")):epsilon`;
+    `(add_ebin a1 b1):epsilon`] thm1) THEN 
+  REWRITE_TAC[
+    ASSUME `proper_nat_construct (add_ebin a1 b1)`;
+    proper_nat_construct] THEN 
+  DISCH_TAC THEN 
+  ASM_REWRITE_TAC[] THEN 
+  IS_EXPR_TYPE_TAC THEN 
+  REPEAT DISCH_TAC THEN 
+  ASM_REWRITE_TAC[] THEN
+  CONSTRUCTION_TO_QUOTE_TAC THEN
+  LAW_OF_DISQUO_TAC THEN  
+  ASM_REWRITE_TAC[add_odd_even] THEN 
+  QUOTE_TO_CONSTRUCTION_TAC THEN 
+  MP_TAC(SPEC `(add_ebin a1 b1)` thm3) THEN 
+  ASM_REWRITE_TAC[] THEN 
+  MP_TAC(fin_x_sub) THEN 
+  ASM_REWRITE_TAC[] THEN 
+  DISCH_TAC THEN
+  ASM_REWRITE_TAC[] THEN 
+  DISCH_TAC THEN 
+  ASM_REWRITE_TAC[] THEN 
+  MP_TAC(SPEC `b1:epsilon` (ASSUME 
+    `!y. proper_nat_construct a1
+    ==> proper_nat_construct y
+    ==> (eval (add_ebin a1 y) to (nat)) =
+    add_unary (eval (a1) to (nat)) (eval (y) to (nat))`)) THEN 
+  REWRITE_TAC[
+    y_quoapp_case1_sub;
+    ASSUME `proper_nat_construct a1`;
+    ASSUME `proper_nat_construct b1`] THEN 
+  DISCH_TAC THEN 
+  ASM_REWRITE_TAC[] THEN 
+  MP_TAC(SYM(SPEC 
+    `add_unary (eval (a1:epsilon) to (nat)) (eval (b1:epsilon) to (nat))` 
+    carry_one)) THEN 
+  DISCH_TAC THEN 
+  ASM_REWRITE_TAC[] THEN 
+  REWRITE_TAC[add_even_odd;one_def;take_s;id_of_plus] THEN
+  REWRITE_TAC[put_s;add_one_even] 
+  ]
+  ;REWRITE_TAC[proper_nat_construct]
+  ]
+  ]
+  ]
+  ;REWRITE_TAC[proper_nat_construct]
+  ]
+  ]
+  ]
+  );;
```

Comment:
That is indeed quite the monster. I definitely think it should be split into lemmas, corresponding to the shape that I indicated about in the comment above, to isolate the shape-driven tactics from the nat-specific parts, as much as possible.